README.md
# RPackUtils

![Package on PYPI](https://img.shields.io/pypi/v/RPackUtils.svg) ![Supported Python versions](https://img.shields.io/pypi/pyversions/RPackUtils.svg)

**Table of Contents**
* [Introduction](#introduction)
* [Requirements](#requirements)
* [Installation](#installation)
* [Configuration](#configuration)
* [Usage](#usage)
    * [rpackcc](#rpackcc)
    * [rpackbioc](#rpackbioc)
    * [rpackmran](#rpackmran)
    * [rpackq](#rpackq)
    * [rpacki](#rpacki)
    * [rpackc](#rpackc)
    * [rpackd](#rpackd)
    * [rpackm](#rpackm)
    * [rpackg](#rpackq)
    * [rpackscan](#rpackscan)
* [Repository types](#repository-types)
* [License checking](#license-checking)
* [Third parties](#third-parties)
* [License](#license)

## Introduction

R package dependencies manager and mirroring tool for *Bioconductor* and
*CRAN* written in Python.  With **RPackUtils** you can:
*  Install R packages in a reproducible manner with the control of packages
   versions and dependencies
*  Clone existing R environments
*  Mirror past and current *CRAN* snapshots published on *MRAN*
*  Mirror past and current *Bioconductor* versions
*  Use Artifactory as a R packages repository
*  Check and scan licenses of R packages

It is still under development.

## Requirements

**RPackUtils** needs Python3.

## Installation

1. Get the source code.
```bash
$ git clone https://github.com/pmpsa-hpc/RPackUtils.git
Cloning into 'RPackUtils'...
[...]
```

2. Enter the root folder and create a Virtualenv.
```bash
$ cd RPackUtils
$ python -m venv .
```

3. Activate the Virtualenv
```bash
$ . bin/activate
(RPackUtils) $
```

4. Upgrade pip and setuptools (recommended)
```bash
$ pip install -U pip
[...]

$ pip install -U setuptools
[...]
```

5. Build the Application
```bash
$ python setup.py build
[...]
```

6. Optionally launch the unit tests
```bash
(RPackUtils) $ cd tests
(RPackUtils) $ pytest -sv [-m "not slow"]
[...]
```
You can optionally unselect tests marked as "slow".

7. Install
```bash
(RPackUtils) $ python setup.py build install
[...]
```
This will install the binary commands in the *bin/* folder of your
Virtualenv.

You may want to specify an installation location with *--prefix*, please
check the documentation of the setup tool by typing :
```bash
(RPackUtils) $ python setup.py install --help
```

## Configuration

Customize the provided configuration file sample.

```bash
RPackUtils/tests/resources $ cp rpackutils.conf ~/
$ vim ~/rpackutils.conf
```

The configuration file has to be passed as an argument for commands requiring it:
```bash
$ RPACKCOMMAND --config ~/rpackutils.conf ...
```

In this configuration file, you have to define your available repositories.
**RPackUtils** currently supports the following repository types:
* JFrog Artifactory with the directive *artifactory_repos*
* Any R environment with *renvironment_repos*
* Any repository on your local file system, like a folder with *local_repos*

```
# [REPO_TYPE]_repos = [list of names]
#
# where each name links to an existing configuration section.
#
[repositories]
artifactory_repos = artifactory, artifactorydev
renvironment_repos = R-3.1.2, R-3.2.5
local_repos = local

[artifactory]
baseurl = https://YOUR_ARTIFACTORY_HOSTNAME/artifactory
user = YOUR_ARTIFACTORY_USER
password = "YOUR_ARTIFACTORY_PASSWORD"
verify = /toto/Certificate_Chain.pem
repos = R-3.1.2, Bioc-3.0, R-local, R-Data-0.1

[artifactorydev]
baseurl = https://YOUR_ARTIFACTORY_DEV_HOSTNAME/artifactory
user = YOUR_ARTIFACTORY_DEV_USER
password = "YOUR_ARTIFACTORY_DEV_PASSWORD"
verify = /toto/Certificate_Chain_Dev.pem
repos = R-3.1.2, Bioc-3.0, R-local, R-Data-0.1

[local]
baseurl = /home/john/RPackUtils/repository
repos = local1, local2

[R-3.1.2]
rhome = /home/john/opt/R-3.1.2
librarypath = lib64/R/library
licensecheck = True

[R-3.2.5]
rhome = /home/john/opt/R-3.2.5
librarypath = lib64/R/library
```

To know more about supported repository types, please read section
[Repository types](#repository-types).

In the *R-3.1.2* environment, we have enabled the *licensecheck*. This
feature is supported by any R environment and is disabled by default. For
more information about this feature, please refer to the dedicated section
[License checking](#license-checking).

To customize the temporary files location you have to change the
corresponding environment variable used by the *tempfile* Python
module. For bash shells, you have to change *TMPDIR*.

```bash
export TMPDIR=/home/john/tmp
```

## Usage

The entry points or commands are installed in *$PREFIX/bin*. This is either
inside your *bin/* folder of your Virtualenv or in the *bin/* folder of the
specified *--prefix*.

| Command                 | Purpose                                                                                              |
|-------------------------|------------------------------------------------------------------------------------------------------|
| [rpackcc](#rpackcc)     | Check the configuration File                                                                         |
| [rpackbioc](#rpackbioc) | Query the Bioconductor repository for available releases                                             |
| [rpackmran](#rpackmran) | Query the MRAN repository for available snapshots                                                    |
| [rpackq](#rpackq)       | Search accross repositories for a package or a list of packages                                      |
| [rpacki](#rpacki)       | Install R packages with resolved dependencies                                                        |
| [rpackd](#rpackd)       | Download R packages and resolved dependencies (dry-run install)                                      |
| [rpackc](#rpackc)       | Install R packages based on an existing environments (clone)                                         |
| [rpackm](#rpackm)       | Download R packages from a specified repository (CRAN, Bioc or Local) and upload them to Artifactory (mirror) |
| [rpackg](#rpackg)       | Generate a dependencies graph                                                                        |
| [rpackscan](#rpackscan) | Scan a repository or an R environment                                                                |

The following sections provide use cases for each command.

### rpackcc

The goal of this command is to parse the configuration file and validate
it. **RPackUtils** will try to connect to any defined remote repository
and check any local file system path exist.

```bash
$ rpackcc -h
usage: rpackcc [-h] --config CONFIG

Check the configuration file and report issues if any

optional arguments:
  -h, --help       show this help message and exit
  --config CONFIG  RPackUtils configuration file
```

Example with a connection issue:
```bash
$ rpackcc --config ~/rpackutils.conf
Checking the configuration file...
Building Artifactory instance "artifactory"
Checking connection to https://artifactory.local/artifactory ...
FATAL: Cannot connect to https://artifactory.local/artifactory: Could not find a suitable TLS CA certificate bundle, invalid path: /toto/Certificate_Chain.pem
(...)
```

Example of a successful validation:
```bash
$ rpackcc --config ~/rpackutils.conf
Building Artifactory instance "myartifactory"
Checking connection to https://myartifactory/artifactory ...
Connection to https://myartifactory/artifactory established.
Building R environment instance "R-3.1.2"
Building R environment instance "R-3.2.2"
Building R environment instance "R-3.2.5"
Building R environment instance "R-3.2.5_1"
Building R environment instance "R-3.2.5_2"
Building Local repository instance "local1"
Building Local repository instance "local2"
Time elapsed: 0.870 seconds.
```

### rpackbioc

Get all available **Bioconductor** releases from their website.

```bash
$ rpackbioc -h
usage: rpackbioc [-h]

Query the Bioconductor repository for available releases

optional arguments:
  -h, --help  show this help message and exit
```

Providing no argument will list all available releases.

```bash
$ rpackbioc
Checking connection to https://www.bioconductor.org ...
Connection to https://www.bioconductor.org established.
Bioconductor 3.7 (devel)
Bioconductor 3.6 (release)
Bioconductor 3.5
Bioconductor 3.4
Bioconductor 3.3
Bioconductor 3.2
Bioconductor 3.1
Bioconductor 3.0
[...]
```

### rpackmran

```bash
$ rpackmran -h
usage: rpackmran [-h] [--Rversion RVERSION] [--procs PROCS]
                 [--dump DUMPFILEPATH] [--restore RESTOREFROMFILEPATH]

Search accross repositories for a package or a list of packages

optional arguments:
  -h, --help            show this help message and exit
  --Rversion RVERSION   The version of R, as "x.y.z" (ignored when --restore
                        is used)
  --procs PROCS         Number of parallel processes used to parse snapshots
                        properties, default=50
  --dump DUMPFILEPATH   a file where to dump the results
  --restore RESTOREFROMFILEPATH
                        a file where to read the results from
```

Providing no argument will fetch all available snapshots for all archived R
versions.

```bash
$ rpackmran 
I will use 50 parallel processes to parse the R version from snapshots dates.
Fetching available MRAN snapshots from the Internet...
Checking connection to https://mran.revolutionanalytics.com ...
Connection to https://mran.revolutionanalytics.com established.

CHANGES.txt
0.2.1
-----

- RPACU-54 Bug fix, custom library path not taken into account while
  installing
- Fix missing column 'linkingTo' for the csv file output (rpackscan)

0.2.0
-----

- RPACU-48 In the configuration file, allow the user to specify both a
  relative and an absolute path to the R library
- RPACU-49 Update the credentials placeholders so that they more
  clearly appear as placeholders
- RPACU-50 Switch the license from GPL v2 to Apache 2.0
- RPACU-51 Check the output messages from stderr and stdout even when
  the R installation command returns a success code
- RPACU-52 Fix DESCRIPTION file parsing, read the "LinkingTo" field

0.1.9
-----

- FIX bug in computing the license class
- Github repository migration from pmpsa-hpc to philipmorrisintl

0.1.8
-----

- Add a new option to rpackm called "--keep" the keep downloaded
  packages
- FIX bug in computing the license class
- FIX invalid escape sequence in regex for the CRAN provider
- FIX DeprecationWarning: The 'warn' method is deprecated for the
  logger class, we have to use 'warning' instead

0.1.7
-----

- RPACU-38 licensecheck boolean is not parsed correctly while reading the
  configuration file
- RPACU-37 new option '--overwrite-specified' to only overwrite specified
  packages and not all as for the option '--overwrite'
- RPACU-35 new command to scan a repository to fetch R packages metadata
- Clean and reformat the source code
- Skip tests requiring any external resource when the later is unavailable
- Add ability to choose a destination folder while mirroring, so the
  downloaded packages are kept there and the mirror command can be launched
  again
- Support a local repository while mirroring
- Document the supported repository types with a class diagram
- The configuration sections ('artifactory_repos', 'renvironment_repos',
  'local_repos') are not mandatory
- RPACU-41 Check the Artifactory client implementation supports sub folders
  as repositories

0.1.6
-----

- Prepare and publish to PYPI

0.1.5
-----

- RPACU-16 Support multiple repository providers
- RPACU-18 Enhance the R environments cloning (rpacks) by reusing temporary
  files
- RPACU-17 reIntegrate the DepTree with the networkx module
- RPACU-19 Write unit tests, use mocks and mark slower tests
- RPACU-20 Update the documentation (README.md)
- RPACU-21 Enhance the DESCRIPTION file parser, write a recursive parser
  (PackInfo)
- RPACU-22 Support license-classes (BLACKLISTED, RESTRICTED, ALLOWED)
  (PackInfo)
- RPACU-23 Enhance version numbers comparison and support bamboo build
  numbers
- RPACU-24 FIX version mismatch in the LooseVersion comparison
- RPACU-25 The nloptr package is downloading external stuff, the subprocess
  module hangs and never complete
- RPACU-26 Ignore binary R packages, report an error and continue to
  process
- RPACU-27 Adapt the configuration file format and write a parser to
  instantiate corresponding objects (repositories)
- RPACU-28 timestamp every task
- RPACU-29 Use a logger to output messages instead of the "print" function
- RPACU-30 Compute the license-class in PackInfo (BLACKLISTED, RESTRICTED,
  ALLOWED)
- RPACU-31 Support imports, depends ans suggests in dependency graphs
- RPACU-33 Dependency Graph: Search across all defined Artifactory
  repositories when --repoparam is not specified
- RPACU-34 Introduce a new license-class "UNKNOWN" and deal with the
  external file references (LICENSE file)

0.1.4
-----

- RPACU-14 Support Bioconductor mirroring, new strategy to fetch packages
  versions
- new command to list BIOC releases 'rpackbioc'

0.1.3
-----

- RPACU-12 implement parallel downloads and Artifactory deployments
- RPACU-12 rpackmran is now way faster thanks to the mutiprocessing
  implementation

0.1.2
-----

- RPACU-11 FIX the cURL command used to deploy to Artifactory

0.1.1
-----

- RPACU-10 Added rpackl function for listing dependencies

0.1.0
-----

- RPACU-9 Get Artifactory SSL Certificate and credentials from a file
- Add the wheel package to the dependencies list to build binary packages

0.0.9
-----

- RPACU-7 Adjust the python module to work with new Artifactory service
  user settings

0.0.8
-----

- RPACU-8 Support SSL while connection to Artifactory

0.0.7
-----

- RPACU-2 Add support to mirror CRAN

0.0.6
-----

- Fix tree method not found anymore in DepsManager

0.0.5
-----

- Fix issue with wrong package name exception

0.0.4
-----

- Fix alphabetical ordering in version number that has to be numerical one.

0.0.3
-----

- Updated default repositories

0.0.2
-----

- Version that is using Artifactory repositories. It has the following

rpackutils/deps.py
###################################################################
# This program is distributed in the hope that it will be useful, #
# but WITHOUT ANY WARRANTY; without even the implied warranty of  #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the    #
# GNU General Public License for more details.                    #
###################################################################

import os
import errno
import requests
import sys
import argparse
import subprocess
import tempfile
import multiprocessing
import time

from .graph import Graph
from .graph import Node
from .repos import RRepository
from .repos import RRepositoryException
from .artifactory import ArtifactoryHelper

ARTIFACTORY_CONFIG_HELP = "File specifying the Artifactroy configuration.\n" \
    "Sample content:\n" \
    "\n" \
    "[global]\n" \
    "artifactory.url = \"https://artifactoryhost/artifactory\"" \
    "artifactory.user = \"artiuser\"" \
    "artifactory.pwd = \"***\"" \
    "artifactory.cert = \"Certificate_Chain.pem\""

base_packages = [
            'R',
            'utils',
            'mgcv',
            'foreign',
            'survival',
            'tcltk',
            'translations',
            'cluster',
            'MASS',
            'grDevices',
            'nlme',
            'grid',
            'base',
            'compiler',
            'KernSmooth',
            'class',
            'spatial',
            'boot',
            'tools',
            'lattice',
            'nnet',
            'graphics',
            'methods',
            'codetools',
            'stats4',
            'rpart',
            'datasets',
            'stats',
            'splines',
            'parallel',
            'Matrix',
        ]

def enum(**enums):
    return type('Enum', (), enums)

def mkdir_p(path):
    try:
        os.makedirs(path)
    except OSError as exc:
        if exc.errno == errno.EEXIST and os.path.isdir(path):
            pass
        else: raise


PackStatus = enum(
        UNKNOWN=-1,
        PARSED=0,
        DOWNLOADED=1,
        DOWNLOAD_FAILED=2,
        INSTALLED=3,
        INSTALLATION_FAILED=4)


class PackNode(Node):
    def __init__(self, idt):
        super(PackNode, self).__init__(idt)
        self._version = None
        self._status = None
        self._packurl = None
        self.reponame = None

    @property
    def packurl(self):
        return self._packurl

    @packurl.setter
    def packurl(self, v):
        self._packurl = v

    @property
    def version(self):
        return self._version

    @version.setter
    def version(self, v):
        self._version = v

    @property
    def status(self):
        return self._status

    @status.setter
    def status(self, v):
        self._status = v

    @property
    def reponame(self):
        return self._reponame

    @reponame.setter
    def reponame(self, v):
        self._reponame = v

    def __str__(self):
        return ('{0} - {1} - {2} - {3}'.format(
            self.idt, self.version, self.status, self.reponame))

class DepsManager(object):
    def __init__(self, artifactoryConfig=None):
        self._repos = {}
        self._processed = []
        self._nofound = []
        self._trace = True
        self._fun = None
        self._funargs = {}
        self._artifactoryConfig = artifactoryConfig

    @property
    def fun(self):
        return self._fun

    @fun.setter
    def fun(self, v):
        self._fun = v

    @property
    def notfound(self):
        return self._nofound

    @property
    def trace(self):
        return self._trace

    @trace.setter
    def trace(self, v):
        self._trace = v

    @property
    def funargs(self):
        return self._funargs

    @funargs.setter
    def funargs(self, v):
        self._funargs = v

    @property
    def repositories(self):
        return self._repos.keys()

    @repositories.setter
    def repositories(self, v):
        for r in v:
            k = r.strip()
            self._repos[k] = RRepository(k, self._artifactoryConfig)

    def repopack(self, p):
        for repo in self._repos:
            if p in self._repos[repo].packages:
                return repo
        return None

    def processnode(self, node):
        if node.idt in base_packages and self._trace:
            print('Package: {0} already installed'.format(node.idt))
            return
        if node.idt in self._nofound:
            if self._trace:
                print('Package: {0} was not found'.format(node.idt))
                return

        if not node.idt in self._processed:
            repo = self.repopack(node.idt)
            if repo is None:
                if self._trace:
                    print('Package: {0} was not found'.format(node.idt))
                self._nofound.append(node.idt)
                return
            _, version, deps = self._repos[repo].description(node.idt)
            node.version = version
            node.packurl = self._repos[repo].packurl(node.idt)
            node.reponame = repo
            deps = [x for x in deps if x not in base_packages]
            
            for dep in deps:
                depnode = PackNode(dep)
                self.processnode(depnode)
                
            print('Processing node: {0}...'.format(node.idt))
            self._fun(node, self._funargs)
            self._processed.append(node.idt)

    @staticmethod
    def applyfun2graph(g, fun=None, *a, **k):
        for idt in g.nodesidts:
            #print('Processing: {0}...'.format(idt))
            n = g.node(idt)
            DepsManager.applyfun2node(n, fun=fun, *a, **k)

    @staticmethod
    def applyfun2node(n, fun=None, *a, **k):
        for childidt in n.connections:
            child = n.node(childidt)
            DepsManager.applyfun2node(child, fun=fun, *a, **k)
        fun(n, *a, **k)

    def tree(self, packages, repos=['R-3.1.2', 'Bioc-3.0', 'R-local'], useref=False):
        g = Graph()
        notfound = []
        rrepos = {}
        for repo in repos:
            rrepos[repo] = RRepository(repo, self._artifactoryConfig)
        for pack in packages:
            print('----------------------------------------------------------')
            print('Processing package: {0}'.format(pack))
            found = False
            if pack in base_packages:
                #print('Ignoring base package: {0}'.format(pack))
                continue
            for repo in repos:
                rrepo = rrepos[repo]
                if pack in rrepo.packages:
                    packname, version, deps = rrepos[repo].description(pack)
                    print('Package: {0}, version: {1} found in repos: {2}'.format(
                        pack, version, repo))
                    packurl = rrepos[repo].packurl(pack)
                    reponame = repo
                    found = True
            if found:
                node = PackNode(packname)
                node.version = version
                node.status = PackStatus.PARSED
                node.packurl = packurl
                node.reponame = reponame
                if not packname in base_packages:
                    g.addnode(node)
                for dep in deps:
                    if not dep in base_packages:
                        if not dep in g.nodesidts:
                            depnode = PackNode(dep)
                            for repo in repos:
                                rrepo = rrepos[repo]
                                if dep in rrepo.packages:
                                    depname, version, _ = rrepos[repo].description(dep)
                                    print('Dep: {0}, version: {1} found in repos: {2}'.format(
                                        dep, version, repo))
                                    packurl = rrepos[repo].packurl(dep)
                            depnode.version = version
                            depnode.packurl = packurl
                            depnode.reponame = reponame
                            depnode.status = PackStatus.PARSED
                            g.addnode(depnode)
                            if not useref:
                                packages.append(dep)
                        g.connect(node.idt, dep, 1)
            else:
                notfound.append(pack)
        return g, notfound

    def download(self, n, args):
        # print('====> Downloading package: {0}'.format(n.idt))
        if args['with_r_cmd']:
            if args['rlibpath'] is not None:
                lib_path_cmd = '--library {0}'.format(args['rlibpath'])
            else:
                lib_path_cmd = ''
            if args['prefix'] is None:
                prefix = ''
            else:
                prefix = args['prefix']
            cmd_filepath = os.path.join(args['dest'], 'install.sh')
            if os.path.exists(cmd_filepath):
                f = open(cmd_filepath, 'a+')
            else:
                f = open(cmd_filepath, 'x')
            cmd = '{0} CMD INSTALL {1} {2}'.format(
                os.path.join(args['rhome'], 'bin', 'R'),
                lib_path_cmd,
                os.path.join(prefix, '{0}_{1}.tar.gz'.format(
                    n.idt, n.version))
            )
            f.write(cmd)
            f.write('\n')
            f.close()
        repo = RRepository(n.reponame)
        filepath = repo.download(n.idt, os.path.join(args['dest'],
                '{0}_{1}.tar.gz'.format(n.idt, n.version)))
        if os.path.exists(filepath):
            n.status = PackStatus.DOWNLOADED
        else:
            n.status = PackStatus.DOWNLOAD_FAILED
            print('Failed to download package: {0}'.format(
                n.idt))
            sys.exit(-1)
        

    def installpack(self, n, args):
        rlibpath = args['rlibpath']
        rhome = args['rhome']
        print('====> Instaling package: {0}'.format(n.idt))
        self._installpack(n, rhome, rlibpath)

    def _installpack(self, n, rhome, rlibpath):
        if rlibpath is not None:
            path = rlibpath
        else:
            path = os.path.join(rhome, 'lib64', 'R', 'library')
        installed = os.listdir(path)
        if n.idt in installed:
            print('Package: {0} already installed'.format(n.idt))
            n.status = PackStatus.INSTALLED
            return
        repo = RRepository(n.reponame, self._artifactoryConfig)
        filepath = repo.download(n.idt)
        if os.path.exists(filepath):
            n.status = PackStatus.DOWNLOADED
        else:
            n.status = PackStatus.DOWNLOAD_FAILED
            print('Failed to download package: {0}'.format(
                n.idt))
            sys.exit(-1)

        cmd = "{0}/bin/R CMD INSTALL {1}".format(
                rhome,
                filepath,
                )
        res = subprocess.call(cmd, shell=True)
        if res != 0:
            n.status = PackStatus.INSTALLATION_FAILED
            print('Installation FAILED for: {0}'.format(n.idt))
        else:
            n.status = PackStatus.INSTALLED
        os.remove(filepath)


def checkrhome(rhome):
    if not os.path.exists(rhome):
        print('No R installation found at: {0}'.format(
            rhome))
        sys.exit(-1)
    if not os.path.exists(os.path.join(rhome, 'bin', 'R')):
        print('Invalid R-home: {0}'.format(
            rhome))
        sys.exit(-1)

def checkrlibpath(path):
    if path is not None:
        print('Installing packages in library: {0}'.format(path))
        if not os.path.exists(path):
            print((
                'Specified location for packages',
                'destination: {0} does not exist').format(path))

def checkrlibpathref(path):
    if path is not None:
        print('Using reference library path: {0}'.format(path))
        if not os.path.exists(path):
            print((
                'Specified location for packages',
                'reference: {0} does not exist').format(path))

def welcome(rhome, repos):
    print('Using R: {0}'.format(rhome))
    print('Using Repositories: {0}'.format(repos))

def tracenotfound(notfound):
    if notfound:
        print('They are package that are not found:')
        #print('\n'.join('{}: {}'.format(*k) for k in enumerate(notfound)))
        print(str(notfound))

def showadditionalactions():
    print('Add HPC certificate to RCurl installation')

def rpacks_install():
    parser = argparse.ArgumentParser(
            description='Install packages based on a list of packages')
    parser.add_argument(
            '--repositories',
            dest='repos',
            action='store',
            default='R-3.1.2,Bioc-3.0,R-local,R-Data-0.1',
            help='Comma separated list of repositories. Default are: R-3.1.2, Bioc-3.0, R-local,R-Data-0.1',
            ) and None
    parser.add_argument(
            '--R-lib-path',
            dest='rlibpath',
            action='store',
            default=None,
            help='Path to the R library path where to install packages',
            ) and None
    parser.add_argument(
            '--R-home',
            dest='rhome',
            action='store',
            default=None,
            help='Path to the R installation to use for installing',
            ) and None
    parser.add_argument(
            '--packages',
            dest='packages',
            action='store',
            default=None,
            required=True,
            help='Comma separated package names to install',
            ) and None
    parser.add_argument(
            '--config',
            dest='artifactoryConfig',
            action='store',
            default=None,
            required=True,
            help=ARTIFACTORY_CONFIG_HELP,
            ) and None
    
    args = parser.parse_args()
    repos = [x.strip() for x in args.repos.split(',')]
    rlibpath = args.rlibpath
    rhome = args.rhome
    packages = [x.strip() for x in args.packages.split(',')]
    artifactoryConfig = args.artifactoryConfig
    # Check some parameters
    checkrhome(rhome)
    checkrlibpath(rlibpath)
    welcome(rhome, repos)

    dm = DepsManager(artifactoryConfig)
    dm.repositories = repos
    dm.fun = dm.installpack
    dm.funargs = {'rlibpath': rlibpath, 'rhome': rhome}
    for pack in packages:
        n = PackNode(pack)
        dm.processnode(n)
    if dm.notfound:
        tracenotfound(dm.notfound)

def rpacks_clone():
    parser = argparse.ArgumentParser(
            description='Install R packages based on an existing environments (clone)')
    parser.add_argument(
            '--repositories',
            dest='repos',
            action='store',
            default='R-3.1.2,Bioc-3.0,R-local,R-Data-0.1',
            help='Comma separated list of repositories. Default are: R-3.1.2, Bioc-3.0, R-local,R-Data-0.1',
            ) and None
    parser.add_argument(
            '--R-lib-path',
            dest='rlibpath',
            action='store',
            default=None,
            help='Path to the R library path where to install packages',
            ) and None
    parser.add_argument(
            '--R-home',
            dest='rhome',
            action='store',
            default=None,
            help='Path to the installation to use',
            ) and None
    parser.add_argument(
            '--R-lib-refpath',
            dest='rlibpathref',
            action='store',
            default=None,
            help='Path to the R library path used as reference',
            ) and None
    parser.add_argument(
            '--config',
            dest='artifactoryConfig',
            action='store',
            default=None,
            required=True,
            help=ARTIFACTORY_CONFIG_HELP,
            ) and None

    args = parser.parse_args()
    repos = [x.strip() for x in args.repos.split(',')]
    rlibpath = args.rlibpath
    rhome = args.rhome
    rlibpathref = args.rlibpathref
    artifactoryConfig = args.artifactoryConfig
    # Check some parameters
    checkrhome(rhome)
    checkrlibpath(rlibpath)
    checkrlibpathref(rlibpathref)
    welcome(rhome, repos)
    packages = os.listdir(rlibpathref)
    packages = [x for x in packages]

    dm = DepsManager(artifactoryConfig)
    dm.repositories = repos
    dm.fun = dm.installpack
    dm.funargs = {'rlibpath': rlibpath, 'rhome': rhome}
    for pack in packages:
        n = PackNode(pack)
        dm.processnode(n)
    if dm.notfound:
        tracenotfound(dm.notfound)

    print('Building dependency tree...')
    g, notfound = dm.tree(packages, repos, useref=True)
    tracenotfound(notfound)
    DepsManager.applyfun2graph(g, fun=dm.installpack,
            rhome=rhome, rlibpath=rlibpath)
    showadditionalactions()

def rpacks_query():
    parser = argparse.ArgumentParser(
            description='Search accross repositories for a package or a list of packages')
    parser.add_argument(
            '--repositories',
            dest='repos',
            action='store',
            default='R-3.1.2,Bioc-3.0,R-local,R-Data-0.1',
            help='Comma separated list of repositories. Default are: R-3.1.2, Bioc-3.0, R-local,R-Data-0.1',
            ) and None
    parser.add_argument(
            '--packages',
            dest='packages',
            action='store',
            default=None,
            required=True,
            help='Comma separated package names to install',
            ) and None
    parser.add_argument(
            '--config',
            dest='artifactoryConfig',
            action='store',
            default=None,
            required=True,
            help=ARTIFACTORY_CONFIG_HELP,
            ) and None
    args = parser.parse_args()
    repos = [x.strip() for x in args.repos.split(',')]
    packages = [x.strip() for x in args.packages.split(',')]
    artifactoryConfig = args.artifactoryConfig
    if not repos:
        print('Repository is not provided')
        sys.exit(-1)
    if not packages:
        print('No package name provided')
    rrepos = {}
    for repo in repos:
        rrepos[repo] = RRepository(repo, artifactoryConfig)
    for pack in packages:
        found = False
        for repo in repos:
            rrepo = rrepos[repo]
            if pack in rrepo.packages:
                found = True
                packname, version, deps = rrepos[repo].description(pack)
                print('-------------------------------------')
                print('Package: {0}  Version: {1} in repository: {2}'.format(
                    pack, version, repo))
                print('Depends on: {0}'.format(str(deps)))
        if not found:
                print('-------------------------------------')
                print('Package: {0} not found in repositories: {1}'.format(
                    pack, str(repos)))

def rpacks_download():
    parser = argparse.ArgumentParser(
            description=('Download R packages and resolved dependencies'))
    parser.add_argument(
            '--repositories',
            dest='repos',
            action='store',
            default='R-3.1.2,Bioc-3.0,R-local,R-Data-0.1',
            help='Comma separated list of repositories. Default are: R-3.1.2, Bioc-3.0, R-local,R-Data-0.1',
            ) and None
    parser.add_argument(
            '--packages',
            dest='packages',
            action='store',
            default=None,
            required=True,
            help='Comma separated package names to install',
            ) and None
    parser.add_argument(
            '--config',
            dest='artifactoryConfig',
            action='store',
            default=None,
            required=True,
            help=ARTIFACTORY_CONFIG_HELP,
            ) and None
    parser.add_argument(
            '--with-R-cmd',
            dest='with_r_cmd',
            action='store_true',
            default=None,
            required=False,
            help='Print to the console the R command to install them',
            ) and None
    parser.add_argument(
            '--R-home',
            dest='rhome',
            action='store',
            default=None,
            help='Path to the installation to use',
            ) and None
    parser.add_argument(
            '--R-lib-path',
            dest='rlibpath',
            action='store',
            default=None,
            required=True,
            help='Path to the R library path where to install packages',
            ) and None
    parser.add_argument(
            '--prefix',
            dest='prefix',
            action='store',
            default=None,
            help='Path to the location where R packages have been downloaded', 
            ) and None
    parser.add_argument(
            '--dest',
            dest='dest',
            action='store',
            default=None,
            help=('Path to the folder where packages are downloaded. If not'
                  'provided, working folder will be used.'), 
            ) and None
    args = parser.parse_args()
    if args.with_r_cmd is not None and args.rhome is None:
        print('You must specify the --R-home when using --with-R-cmd')
        sys.exit(-1)
    repos = [x.strip() for x in args.repos.split(',')]
    packages = [x.strip() for x in args.packages.split(',')]
    dest = args.dest
    if args.dest is None:
        dest = os.getcwd()
    artifactoryConfig = args.artifactoryConfig
    if not repos:
        print('Repository is not provided')
        sys.exit(-1)
    dm = DepsManager(artifactoryConfig)
    dm.repositories = repos
    dm.fun = dm.download
    dm.funargs = {'rlibpath': args.rlibpath, 'rhome': args.rhome,
                  'with_r_cmd': args.with_r_cmd, 'prefix': args.prefix,
                  'dest': dest}
    starttime = time.time()
    for pack in packages:
        node = PackNode(pack)
        dm.processnode(node)
    endtime = time.time()
    timeelapsed = int(endtime - starttime)
    print('Time elapsed: {0} seconds.'.format(timeelapsed))
    if dm.notfound:
        tracenotfound(dm.notfound)
